common/quotas/global/rpc/client.go

```diff
@@ -0,0 +1,191 @@
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"go.uber.org/yarpc"
+	"golang.org/x/exp/maps"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/uber/cadence/client/history"
+	"github.com/uber/cadence/common/log"
+	"github.com/uber/cadence/common/log/tag"
+	"github.com/uber/cadence/common/metrics"
+	"github.com/uber/cadence/common/types"
+)
+
+type (
+	Calls struct {
+		Allowed  int
+		Rejected int
+	}
+)
+
+type (
+	// UpdateResult holds all successes and errors encountered, to be distinct
+	// from fatal errors that imply unusable results (i.e. a normal error return).
+	UpdateResult struct {
+		// all successfully-returned weights.
+		// this may be populated and is safe to use even if Errors is present,
+		// as some shards may have succeeded.
+		Weights map[string]float64
+		// Any unexpected errors encountered, or nil if none.
+		// Weights is valid even if this is present, though it may be empty.
+		//
+		// This should always be nil aside from major programming or rollout
+		// errors, e.g. failures to serialize or deserialize data (coding errors
+		// or incompatible server versions).
+		Err error
+	}
+
+	Client interface {
+		// Update performs concurrent calls to all aggregating peers to send load info
+		// and retrieve new per-key weights from the rest of the cluster.
+		// This is intended to be called periodically in the background per process,
+		// not synchronously or concurrently (per set of keys anyway), as it is fairly
+		// high cost to shard keys and send so many requests.
+		//
+		// Currently, there are no truly fatal errors so this API does not return `error`.
+		// Even in the presence of errors, some successful data may have been loaded,
+		// and that will be part of the UpdateResult struct.
+		Update(ctx context.Context, period time.Duration, load map[string]Calls) UpdateResult
+	}
+
+	client struct {
+		history  history.Client
+		resolver history.PeerResolver
+		thisHost string
+
+		logger log.Logger
+		scope  metrics.Scope
+	}
+)
+
+var _ Client = (*client)(nil)
+
+func New(
+	historyClient history.Client,
+	resolver history.PeerResolver,
+	logger log.Logger,
+	met metrics.Client,
+) Client {
+	return &client{
+		history:  historyClient,
+		resolver: resolver,
+		thisHost: uuid.NewString(), // TODO: would descriptive be better?  but it works, unique ensures correctness.
+		logger:   logger,
+		scope:    met.Scope(metrics.FrontendGlobalRatelimiter),
+	}
+}
+
+func (c *client) Update(ctx context.Context, period time.Duration, load map[string]Calls) UpdateResult {
+	batches, err := c.resolver.GlobalRatelimitPeers(maps.Keys(load))
+	if err != nil {
+		// should only happen if peers are unavailable, individual requests are handled other ways
+		return UpdateResult{
+			Err: fmt.Errorf("unable to shard ratelimit update data: %w", err),
+		}
+	}
+
+	var mut sync.Mutex
+	weights := make(map[string]float64, len(load)) // should get back most or all keys requested
```

Comment:
Is there a risk that partial results will return as zeros because you're initializing the whole set?

Reply:
I think you might be thinking of slice behavior, where `make([]string, len(load))` would leave empty strings in un-set indexes?
maps just reserve space internally to reduce re-allocation, there's no behavioral difference at all between an empty or pre-sized map (unlike slices, bleh).